aiida/calculations/transfer.py

```diff
@@ -0,0 +1,211 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Implementation of CalcJobNode to add two numbers for testing and demonstration purposes."""
+import os
+from aiida import orm
+from aiida.engine import CalcJob
+from aiida.common.datastructures import CalcInfo
+
+
+def validate_transfer_inputs(inputs, ctx):  # pylint: disable=inconsistent-return-statements, unused-argument
+    """Validate the inputs"""
+    source_nodes = inputs['source_nodes']
+    instructions = inputs['instructions']
+
+    if not 'retrieve_files' in instructions.get_dict():
```

Comment:
I would put the validation of the `instructions` that does not depend on `source_nodes` in its own validator and set that directly on the `spec.inputs.instructions` port. The validators of ports within a port namespace are always called before the validator of the namespace itself. That means you can safely assume that if you get to the `inputs` validator, the `instructions` validator has passed and you can assume all the required keys are there.